[flang] Represent use statement in fir.

abidh opened the pull request:
We have a longstanding issue in debug info that use statement is not fully respected. The problem has been described in https://github.com/llvm/llvm-project/issues/160923. This is first part of the effort to address this issue. This PR adds infrastructure to emit `use` statement information in FIR, which will be used by subsequent patches to generate DWARF debug information.

The information about use statement is collected during semantic analysis and stored in `PreservedUseStmt` objects. During lowering, `fir.use_stmt` operations are emitted for each `PreservedUseStmt` object. The `fir.use_stmt` operation captures the module name, `only` list symbols, and any renames specified in the use statement. The `fir.use_stmt` is removed during `CodeGen`.

abidh committed:
[flang] Represent use statement in fir.

This patch adds infrastructure to emit Fortran USE statement
information in FIR, which will be used by subsequent patches to generate
DWARF debug information.

The information about use statement is collected during semantic
analysis and stored in PreservedUseStmt objects. During
lowering, fir.use_stmt operations are emitted for each PreservedUseStmt
object. The fir.use_stmt operation captures the module name, ONLY list
symbols, and any renames specified in the USE statement.

jeanPerier reviewed (commented):
Hi @abidh, thanks for working on this.

I wonder if this representation should better be outside of the func.func.

There is a case that is not addressed here I think (at least in the test), which are the indirect accesses when a module itself has use statements.
In some modern application, there may be many of these indirect accesses, and putting all the use as operation in the func.func would pollute the IR quite a lot.

ergawy reviewed (commented):
Thanks Abid. Just a few comments from side.

abidh committed:
Handle review comments.

1. Removed enum Kind.
2. Simplified the code accordingly.
3. Removed some extra empty lines from the testcase.

ergawy reviewed (commented):
Thanks for the changes Abid. Just one more question.

abidh committed:
Handle review comments (2).

Move handling of parser::UseStmt in PFTBuild and make this processing
conditional on debug information being actually required.

Updated testcase to have only with renames test. Also renamed testcase.

ergawy reviewed (approved):
LGTM! Thanks.

jeanPerier reviewed (approved):
Thanks for addressing my comments, looks great.

abidh committed:
Handle review comments (3).

1. Remove a redundant comment.
2. Use -emit-hlfir instead of -emit-fir.

abidh commented on flang/include/flang/Semantics/scope.h:
```diff
@@ -55,6 +55,16 @@ struct EquivalenceObject {
 using EquivalenceSet = std::vector<EquivalenceObject>;
 
+// Preserved USE statement information for debug info generation.
+// Preserved USE statement information for debug info generation
```
Thanks for catching this. Fixed.

abidh commented:
> After this patch, specifying `-g` now causes a "not yet implemented" error in our tests. Is this the expected behavior, or is it planned to be fixed in another patch? I would appreciate it if you could take a look.

Thanks for reporting it. This is not intended behavior. I will investigate.

abidh commented:
@s-watanabe314 I opened https://github.com/llvm/llvm-project/pull/174387 to fix this. I think all the failed tests should pass now. But it would be great if you could verify with that PR. Thanks.
